core/node/provider.go

```diff
@@ -223,30 +204,71 @@ func newProvidingStrategy(strategy string) interface{} {
 		OfflineIPLDFetcher   fetcher.Factory `name:"offlineIpldFetcher"`
 		OfflineUnixFSFetcher fetcher.Factory `name:"offlineUnixfsFetcher"`
 		MFSRoot              *mfs.Root
+		Provider             provider.System
+		Repo                 repo.Repo
+	}
+	type output struct {
+		fx.Out
+		ProvidingStrategy    string `name:"providingStrategy"`
+		ProvidingKeyChanFunc provider.KeyChanFunc
 	}
-	return func(in input) provider.KeyChanFunc {
+	return func(in input) output {
+		var kcf provider.KeyChanFunc
+
 		switch strategy {
 		case "roots":
-			return provider.NewBufferedProvider(provider.NewPinnedProvider(true, in.Pinner, in.OfflineIPLDFetcher))
+			kcf = provider.NewBufferedProvider(dspinner.NewPinnedProvider(true, in.Pinner, in.OfflineIPLDFetcher))
 		case "pinned":
-			return provider.NewBufferedProvider(provider.NewPinnedProvider(false, in.Pinner, in.OfflineIPLDFetcher))
+			kcf = provider.NewBufferedProvider(dspinner.NewPinnedProvider(false, in.Pinner, in.OfflineIPLDFetcher))
 		case "pinned+mfs":
-			return provider.NewPrioritizedProvider(
-				provider.NewBufferedProvider(provider.NewPinnedProvider(false, in.Pinner, in.OfflineIPLDFetcher)),
+			kcf = provider.NewPrioritizedProvider(
+				provider.NewBufferedProvider(dspinner.NewPinnedProvider(false, in.Pinner, in.OfflineIPLDFetcher)),
 				mfsProvider(in.MFSRoot, in.OfflineUnixFSFetcher),
 			)
 		case "mfs":
-			return mfsProvider(in.MFSRoot, in.OfflineUnixFSFetcher)
+			kcf = mfsProvider(in.MFSRoot, in.OfflineUnixFSFetcher)
 		case "flat":
-			return provider.NewBlockstoreProvider(in.Blockstore)
+			kcf = in.Blockstore.AllKeysChan
 		default: // "all", ""
-			return provider.NewPrioritizedProvider(
+			kcf = provider.NewPrioritizedProvider(
 				provider.NewPrioritizedProvider(
-					provider.NewBufferedProvider(provider.NewPinnedProvider(true, in.Pinner, in.OfflineIPLDFetcher)),
+					provider.NewBufferedProvider(dspinner.NewPinnedProvider(true, in.Pinner, in.OfflineIPLDFetcher)),
 					mfsRootProvider(in.MFSRoot),
 				),
-				provider.NewBlockstoreProvider(in.Blockstore),
+				in.Blockstore.AllKeysChan,
 			)
 		}
+
+		in.Provider.SetKeyProvider(kcf)
+
+		var strategyChanged bool
```

Comment:
The strategy-change detection has been moved here. No changes in code blocks below.